### PR TITLE
Two step source separation

### PR DESCRIPTION
**Pull request by etzinis**
Integrating the two-step source separation in asteroid.

### CODE REVIEW

File: asteroid/filterbanks/simple_adaptive.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    def __init__(self, freq_res, sample_res, n_sources):
+        super().__init__()
+        self.deconv = nn.ConvTranspose1d(n_sources * freq_res,
+                                         n_sources,
+                                         sample_res,
+                                         padding=sample_res // 2,
+                                         stride=sample_res // 2,
+                                         groups=n_sources,
+                                         output_padding=(sample_res // 2) - 1)
```
Comment (mpariente) [inline asteroid/filterbanks/simple_adaptive.py]: This makes one decoder per source, I don't think this is what we want. 
Please have a look at `FreeFB`, we could add `padding` and `output_padding` to `Encoder` and `Decoder` to cover your intended usage, it would be a good addition :smiley:

File: asteroid/filterbanks/simple_adaptive.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class AdaptiveEncoder1D(nn.Module):
```
Comment (mpariente) [inline asteroid/filterbanks/simple_adaptive.py]: `Encoder(FreeFB(*))` does this and more (only padding is missing)

File: asteroid/masknn/blocks.py [hunk 3]
```diff
@@ -50,6 +50,50 @@ def forward(self, x):
...
+        in_chan (int): Number of input channels.
+        hid_chan (int): Number of hidden channels in the depth-wise
+            convolution.
+        kernel_size (int): Size of the depth-wise convolutional kernel.
+        dilation (int): Dilation of the depth-wise convolution.
+
+    References:
+        [1]: "Conv-TasNet: Surpassing ideal time-frequency magnitude masking
+             for speech separation" TASLP 2019 Yi Luo, Nima Mesgarani
+             https://arxiv.org/abs/1809.07454
```
Comment (mpariente) [inline asteroid/masknn/blocks.py]: These blocks were introduced in the v1, we can note that without skip connection it corresponds to them
https://arxiv.org/abs/1809.07454v1

File: asteroid/masknn/blocks.py [hunk 4]
```diff
@@ -50,6 +50,50 @@ def forward(self, x):
         return res_out, skip_out
 
 
+class SeparableDilatedConv2DBlock(nn.Module):
```
Comment (mpariente) [inline asteroid/masknn/blocks.py]: I'd rather have one `Conv1DBlock` which supports with and without skip connections. 
And have `TDConvNet` support this argument as well.
Does this sound feasible?

File: asteroid/masknn/blocks.py [hunk 5]
```diff
@@ -460,3 +504,114 @@ def forward(self, input_data):
         mask_out = self.mask_layer(out)
         mask_out = mask_out.view(batches, self.n_src, self.input_dim, seq_cnt)
         return projection_final, mask_out
+
+
+class TwoStepTDCN(nn.Module):
```
Comment (mpariente) [inline asteroid/masknn/blocks.py]: Architectures in `blocks.py` do not include encoders and decoders, it makes everything inside the the source code more modular and reusable.

File: egs/wham/TwoStep/train.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    train_set = WhamDataset(conf['data']['train_dir'], conf['data']['task'],
+                            sample_rate=conf['data']['sample_rate'],
+                            nondefault_nsrc=conf['data']['nondefault_nsrc'],
+                            normalize_audio=True)
+    val_set = WhamDataset(conf['data']['valid_dir'], conf['data']['task'],
+                          sample_rate=conf['data']['sample_rate'],
+                          nondefault_nsrc=conf['data']['nondefault_nsrc'],
+                          normalize_audio=True)
+
+    if (not train_part == 'filterbank') and (not train_part == 'separator'):
```
Comment (mpariente) [inline egs/wham/TwoStep/train.py]: `if train_part not in ['filterbank', 'separator']:`

File: egs/wham/TwoStep/model.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    def __init__(self, pretrained_filterbank, conf):
+        super().__init__()
+        self.pretrained_filterbank = pretrained_filterbank
+        self.separator = TwoStepTDCN(
+            pretrained_filterbank,
+            bn_chan=conf['masknet']['bn_chan'],
+            hid_chan=conf['masknet']['hid_chan'],
+            kernel_size=conf['masknet']['conv_kernel_size'],
+            n_blocks=conf['masknet']['n_blocks'],
+            n_repeats=conf['masknet']['n_repeats'],
```
Comment (mpariente) [inline egs/wham/TwoStep/model.py]: You can actually feed `n_sources=conf['masknet']['n_src'], **conf['masknet']`

File: egs/wham/TwoStep/model.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        return self.separator.infer_source_signals(x.unsqueeze(1))
+
+
+def get_encoded_paths(conf, train_part=None):
+    exp_dir = conf['main_args']['exp_dir']
+    N = conf['filterbank']['n_filters']
+    L = conf['filterbank']['kernel_size']
+    checkpoint_dir = os.path.join(
+        exp_dir, train_part, 'N_{}_L_{}_checkpoints/'.format(N, L))
+    return exp_dir, checkpoint_dir
```
Comment (mpariente) [inline egs/wham/TwoStep/model.py]: If several _different_ models are trained, they should have different `tag` from bash, and correspond to different experiment folders rather than having different checkpoint folders IMO.

File: egs/wham/TwoStep/model.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+
+def load_best_separator_if_available(conf):
+    filterbank = load_best_filterbank_if_available(conf)
+    _, f_checkpoint_dir = get_encoded_paths(conf, 'filterbank')
+    if filterbank is None:
+        raise ImportError('There are no available filterbanks under: {}\n'
```
Comment (mpariente) [inline egs/wham/TwoStep/model.py]: Rather `FileNotFoundError`, same at line 73

File: egs/wham/TwoStep/model.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+            freq_res=conf['filterbank']['n_filters'],
+            sample_res=conf['filterbank']['kernel_size'],
+            n_sources=conf['masknet']['n_src'])
+    elif model_part == 'separator':
+        if pretrained_filterbank is None:
+            raise ValueError('A pretrained filterbank is required for the '
+                             'initialization of the separator.')
+        model = Model(pretrained_filterbank, conf)
+    else:
+        raise NotImplementedError('Part to train: {} is not '
```
Comment (mpariente) [inline egs/wham/TwoStep/model.py]: Rather use `ValueError`. 
`NotmplementedError` are for abstract methods

File: egs/wham/TwoStep/train.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+def main(conf):
+    filterbank = load_best_filterbank_if_available(conf)
+    _, checkpoint_dir = get_encoded_paths(conf, 'filterbank')
+    if filterbank is None:
+        print('There are no available filterbanks under: {}. Going to '
+              'training.'.format(checkpoint_dir))
+        train_model_part(conf, train_part='filterbank')
+    else:
+        print('Found available filterbank at: {}'.format(checkpoint_dir))
+        inp = input('Do you want to refine it further? y/n\n')
```
Comment (mpariente) [inline egs/wham/TwoStep/train.py]: Cannot take input, these scripts can be ran through job schedulers like slurm or SGE, having an input will just kill your job. Input with timeout would be an option but not so simple to set up.

File: asteroid/masknn/blocks.py (hunk 4 above)
Comment (etzinis) [reply asteroid/masknn/blocks.py:53]: As requested and discussed I will take it away from your core codebase.

File: asteroid/engine/system.py [hunk 12]
```diff
@@ -52,24 +51,29 @@ def forward(self, *args, **kwargs):
         """
         return self.model(*args, **kwargs)
 
-    def common_step(self, batch, batch_nb):
+    def common_step(self, batch, batch_nb, train=True):
```
Comment (mpariente) [inline asteroid/engine/system.py:54]: I guess you cherry-picked three of my commits to have the new `System`.
This is bad habit for two reasons 
- The changes will be asociated with this PR (wrong history)
- If you make a small change on this new version of the file, it is not as obvious to see
There are two ways to deal with this
- Rebase master and force-push
- Merge master into your branch
Because we squash PRs before merging, I prefer the second method.

Comment (etzinis) [reply asteroid/engine/system.py:54]: Oh I am sorry man this was a mistake
